### PR TITLE
Use trees in selects when hierarchy is present

### PR DESCRIPTION
**Pull request by ptbrowne**
## How to test

- Use bathing water with hierarchies (check draft when searching)
- In the data filters, the "observation location" select should display the hierarchy as a tree
- When activating interactive filters, the interactive filter also displays the hierarchy as a tree

## TODO

- [x] When you first open the select, it loads the hierarchy and then closes the select, and you have to open it again

### CODE REVIEW

File: app/components/select-tree.tsx [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+    setTimeout(() => {
+      menuRef.current?.updatePosition();
+    }, 500);
+    setTimeout(() => {
+      menuRef.current?.updatePosition();
+    }, 1000);
+    setTimeout(() => {
+      menuRef.current?.updatePosition();
+    }, 2000);
+  };
```
Comment (bprusinowski) [inline app/components/select-tree.tsx:278]: Why such combo? Wouldn't just first timeout do the job? Or is the transition duration dependent on the number of children in a node?

File: app/components/select-tree.tsx [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+        TransitionProps={transitionProps}
+      >
+        <TreeView
+          defaultSelected={value}
+          defaultExpanded={defaultExpanded}
+          onNodeSelect={handleNodeSelect}
+          onNodeToggle={handleNodeToggle}
+          defaultCollapseIcon={<ExpandMoreIcon />}
+          defaultExpandIcon={<ChevronRightIcon />}
+          // onNodeSelect={handleNodeSelect}
```
Comment (bprusinowski) [inline app/components/select-tree.tsx:341]: 👁️

File: app/docs/form.docs.tsx [hunk 3]
```diff
@@ -28,6 +29,49 @@ const SwitchExample = ({ initialChecked }: { initialChecked?: boolean }) => {
...
+          label: "Numbers",
+          value: "numbers",
+          selectable: false,
+          children: numberOptions,
+        },
+      ]}
+    />
+  );
+};
+
```
Comment (bprusinowski) [inline app/docs/form.docs.tsx:74]: 👌

File: app/charts/shared/chart-data-filters.tsx [hunk 4]
```diff
@@ -255,14 +269,63 @@ const DataFilterBaseDimension = ({
...
+    }
+    if (!isKeyDimension) {
+      opts.push({
+        value: FIELD_VALUE_NONE,
+        label: noneLabel,
+        isNoneValue: true,
+      });
+    }
+    return opts;
+  }, [hierarchy, isKeyDimension, dimensionValues, noneLabel]);
```
Comment (bprusinowski) [inline app/charts/shared/chart-data-filters.tsx:316]: Did not check this as I'm not aware of optional hierarchy in any of the datasets, but wouldn't this append the "None" option at the end of the array (I think it should be at top)?

File: app/components/select-tree.tsx (hunk 1 above)
Comment (ptbrowne) [reply app/components/select-tree.tsx]: Refactored away by using the onEntered / onExited transition props of the tree item transitions.

File: app/charts/shared/chart-data-filters.tsx (hunk 4 above)
Comment (ptbrowne) [reply app/charts/shared/chart-data-filters.tsx]: Ah yes ! I replaced push by unshift :)

**Comment by ptbrowne** [comment]
1. Yes I need to see how to still lazy load the hierarchies (we had gained a significant boost by postponing the downloading of hierarchies), but it should be better done, I need to do refactors here to correctly provide the loading to the select.
2. Fixed, the method may seem hacky (via the --depth CSS variable), but I tried first to use calc(--depth) + 1, but it does not seem to work, it seems we can't have recursion in CSS variables :/
3. I tried to add an indication, I think it's better now, what do you think ? (cc @AnninaWalker )
4. Since I changed the way of doing the padding for the hover effect, this has changed I think. I personally think this looks good already ? 

image

**Review by bprusinowski — APPROVED** [review]
Yes, much better now 💯

**Pull request merged**